File: chart/server/spectralsequence_chart/chart_elements/chart_edge.py
from message_passing_tree.prelude import arguments
from uuid import uuid4
from .. import utils

class ChartEdge:
    def __init__(self, sseq, edge_type, *, source, target, **kwargs):
        self._sseq = sseq
        self.type = edge_type
        if "uuid" in kwargs:
            self.uuid = kwargs["uuid"]
        else:
            self.uuid = str(uuid4())

        self.source = source
        self.target = target

        setattr(self, utils.PROPERTY_PREFIX + "source",  self.source.uuid)
        setattr(self, utils.PROPERTY_PREFIX + "target",  self.target.uuid)

        [message_args, message_kwargs] = arguments(
            type = self.type,
            uuid = self.uuid,
            source = source.uuid,
            target = target.uuid,
            **kwargs
        )
        if "page" in kwargs:
            message_kwargs["page"] = kwargs["page"]
        self._sseq.add_batched_message(self.uuid, "chart.edge.add", message_args, message_kwargs)
        utils.copy_fields_from_kwargs(self, kwargs)

        if "visible" not in kwargs:
            self.visible = True

        self.source._edges.append(self)
        self.target._edges.append(self)

    def needs_update(self):
        self._sseq.add_edge_to_update(self)

    @utils.sseq_property
    def color(self, storage_name):
        self.needs_update()

    @utils.sseq_property
    def bend(self, storage_name):
        self.needs_update()

    def replace_source(self, **kwargs):
        self.source.replace(**kwargs)
    
    def replace_target(self, **kwargs):
        self.target.replace(**kwargs)

    def delete(self):
        del self._sseq.edges[self.uuid]
        # del e._source.edges[e]
        # del e._target.edges[e]

    @staticmethod
    def from_json(sseq, json):
        edge_type = json["type"]
        json["source"] = sseq.classes[json["source"]]
        json["target"] = sseq.classes[json["target"]]
        if edge_type == ChartDifferential.__name__:
            return ChartDifferential(sseq, json.pop("page"), **json)
        if edge_type == ChartStructline.__name__:
            return ChartStructline(sseq, **json)
        if edge_type == ChartExtension.__name__:
            return ChartExtension(sseq, **json)

    def to_json(self):
        return utils.public_fields(self)

class ChartDifferential(ChartEdge):
    def __init__(self, sseq, page, **kwargs):
        super().__init__(sseq, ChartDifferential.__name__, **kwargs, page = page)
        self.page = page

class ChartStructline(ChartEdge):
    def __init__(self, sseq, **kwargs):
        super().__init__(sseq, ChartStructline.__name__, **kwargs)

class ChartExtension(ChartEdge):
    def __init__(self, sseq, **kwargs):
        super().__init__(sseq, ChartExtension.__name__, **kwargs)

File: chart/client/src/sseq/ChartEdge.js
import * as utils from "./utils.js";
import { ChartClass } from "./ChartClass.js";
import { INFINITY } from "../infinity.js";

export class ChartEdge {
    constructor(type, kwargs) {
        this.type = type;
        if(!("source" in kwargs)){
            throw Error(`Edge is missing argument "source".`);
        }
        if(!("target" in kwargs)){
            throw Error(`Edge is missing argument "target".`);
        }
        if(!("visible" in kwargs)) {
            this.visible = true;
        }
        Object.assign(this, kwargs);
        // if(typeof(this.source) !== ChartClass) {
        //     throw Error(`Invalid source of edge : ${JSON.stringify(this.source)}`);
        // }
        // if(typeof(this.target) !== ChartClass) {
        //     throw Error(`Invalid target of edge: ${JSON.stringify(this.target)}`);
        // }
        this._source = this.source;
        this._target = this.target;
        this.source = this.source.uuid;
        this.target = this.target.uuid;
        // utils.assign_fields(this, kwargs, [
        //     { "type" : "mandatory", "field" : "source"},
        //     { "type" : "mandatory", "field" : "target"},
        //     { "type" : "default", "field" : "visible", "default" : true},
        //     { "type" : "optional", "field" : "color"},
        //     { "type" : "optional", "field" : "opacity"},
        //     { "type" : "optional", "field" : "bend"},
        //     { "type" : "optional", "field" : "control_points"},
        //     { "type" : "optional", "field" : "arrow_type"},
        // ])
    }

    _drawOnPageQ(pageRange){
        let max_page = this.max_page || INFINITY;
        let min_page = this.min_page || 0;
        return pageRange[0] <= max_page && min_page <= pageRange[0];
    }

    toJSON() {
        return utils.public_fields(this);
    }
}

export class ChartDifferential extends ChartEdge {
    constructor(kwargs){
        super("differential", kwargs);
        utils.assign_kwarg_mandatory(this, kwargs, "page");
    }

    _drawOnPageQ(pageRange){
        return pageRange[0] === 0 || (pageRange[0] <= this.page && this.page <= pageRange[1]);
    }    
}

export class ChartStructline extends ChartEdge {
    constructor(kwargs){
        super("structline", kwargs);
        if(this.max_page === undefined) {
            this.max_page = INFINITY;
        }
        if(this.min_page === undefined) {
            this.min_page = 0;
        }        
    }

    _drawOnPageQ(pageRange){
        return pageRange[0] <= this.max_page && this.min_page <= pageRange[0];
    }
}

export class ChartExtension extends ChartEdge {
    constructor(kwargs){
        super("extension", kwargs);
    }

    _drawOnPageQ(pageRange){
        return pageRange[0] === INFINITY;
    }

}